### PR TITLE
Create Project Folder Structure: support list of str for leaf folders

### PR DESCRIPTION
**Pull request by BigRoy**
## Changelog Description

Support `list[str]` for leaf entries in `core/project_folder_structure` settings to define folder names instead of requiring dicts with empty values

## Additional info

Should implement and allow what @MustafaJafar did [here](https://github.com/ynput/ayon-core/pull/679#pullrequestreview-2348525202).

Would allow this:
```json
{
    "__project_root__": {
        "Assets": {
            "Characters": {
                "NewRobo": {
                    "work": [
                        "cfx",
                        "concept",
                        "modeling",
                        "walkcycle"
                    ]
                },
                "RoboPoro": {
                    "work": [
                        "cfx",
                        "concept",
                        "modeling",
                        "walkcycle"
                    ]
                },
                "RoboMad": {
                    "work": [
                        "cfx",
                        "concept",
                        "modeling",
                        "walkcycle"
                    ]
                }
            }
        }
    }
}
```

Instead of requiring:
```json
{
    "__project_root__": {
        "Assets": {
            "Characters": {
                "NewRobo": {
                    "work": {
                        "cfx": {},
                        "concept": {},
                        "modeling": {},
                        "walkcycle": {},
                    }
                },
                "RoboPoro": {
                    "work": {
                        "cfx": {},
                        "concept": {},
                        "modeling": {},
                        "walkcycle": {},
                    }
                },
                "RoboMad": {
                    "work": {
                        "cfx": {},
                        "concept": {},
                        "modeling": {},
                        "walkcycle": {},
                    }
                }
            }
        }
    }
}
```

## Testing notes:

1. Setup `ayon+settings://core/project_folder_structure` to include leaf (end) folders using just a list of strings.
2. Creating folder structure should work.

### CODE REVIEW

**Comment by alainxi** [comment]
Does not work for me, when tested with https://github.com/ynput/ayon-core/pull/679
It only creates the first ones :
NewRobo is created, but not RoboPoro and RoboMad
cfx is creared, but not concept , modeling, walkcycle


Also, I prefer the old method (dicts with empty values), but it's still supported, so I'm happy.
The old method makes possible to use a very simple script to create a json from an existing disk folder tree :

```
import json
import os
def generate_json_tree(path):
    result = {}
    for item in os.listdir(path):
        item_path = os.path.join(path, item)
        if os.path.isdir(item_path):
            result[item] = generate_json_tree(item_path)
        else:
            pass
    return result

root_path = "d:/test/forTreeToJsonConvert"
json_tree = generate_json_tree(root_path)

with open("D:/test/folder_tree.json", "w") as f:
    json.dump(json_tree, f, indent=4)
```


The support of `list[str]` for leaf entries makes the script a lot more convoluted, and creates edge cases.

**Comment by MustafaJafar** [comment]
Thank you! 
So, I revisited my created folders and I didn't notice the last time it created the folders in strange way. 
the list of folders are nested folders. 
image

**Comment by BigRoy** [comment]
Thanks - will double check what's up. How interesting to see everyone seems to jump on this one this quickly :D haha

**Comment by BigRoy** [comment]
There's no means to deprecated the old @alainxi so that should continue to work.

@MustafaJafar try again - it should now work with https://github.com/ynput/ayon-core/pull/1169/commits/7baf208c61f992a30de62f29a3aef484207ac542

**Comment by MustafaJafar** [comment]
I got this printed in the console window but no folders were created. 
```json
[
    [
        "__project_root__",
        "Assets",
        "Characters",
        "NewRobo",
        "work",
        "cfx"
    ],
    [
        "__project_root__",
        "Assets",
        "Characters",
        "NewRobo",
        "work",
        "concept"
    ],
    [
        "__project_root__",
        "Assets",
        "Characters",
        "NewRobo",
        "work",
        "modeling"
    ],
    [
        "__project_root__",
        "Assets",
        "Characters",
        "NewRobo",
        "work",
        "walkcycle"
    ],
    [
        "__project_root__",
        "Assets",
        "Characters",
        "RoboPoro",
        "work",
        "cfx"
    ],
    [
        "__project_root__",
        "Assets",
        "Characters",
        "RoboPoro",
        "work",
        "concept"
    ],
    [
        "__project_root__",
        "Assets",
        "Characters",
        "RoboPoro",
        "work",
        "modeling"
    ],
    [
        "__project_root__",
        "Assets",
        "Characters",
        "RoboPoro",
        "work",
        "walkcycle"
    ],
    [
        "__project_root__",
        "Assets",
        "Characters",
        "RoboMad",
        "work",
        "cfx"
    ],
    [
        "__project_root__",
        "Assets",
        "Characters",
        "RoboMad",
        "work",
        "concept"
    ],
    [
        "__project_root__",
        "Assets",
        "Characters",
        "RoboMad",
        "work",
        "modeling"
    ],
    [
        "__project_root__",
        "Assets",
        "Characters",
        "RoboMad",
        "work",
        "walkcycle"
    ]
]
```

File: client/ayon_core/pipeline/project_folders.py [hunk 1]
```diff
@@ -109,4 +119,6 @@ def get_project_basic_paths(project_name):
 
     if isinstance(folder_structure, str):
         folder_structure = json.loads(folder_structure)
-    return _list_path_items(folder_structure)
+    result = _list_path_items(folder_structure)
+    print(result)
+    return []
```
Comment (MustafaJafar) [inline client/ayon_core/pipeline/project_folders.py:124]: 🤔 
```suggestion
    return result
```

**Review by MustafaJafar — APPROVED** [review]
It works 
image

File: client/ayon_core/pipeline/project_folders.py (hunk 1 above)
Comment (MustafaJafar) [reply client/ayon_core/pipeline/project_folders.py]: https://github.com/ynput/ayon-core/pull/1169/commits/1b4f6bea20b2c6e7abac4eb5b57b9377ae78db90

File: client/ayon_core/pipeline/project_folders.py [hunk 2]
```diff
@@ -80,8 +81,17 @@ def create_project_folders(project_name, basic_paths=None):
...
 
-def _list_path_items(folder_structure):
+def _list_path_items(
+        folder_structure: Union[Dict[str, Any], List[str]]):
     output = []
+
+    # Allow leaf folders of the `project_folder_structure` to use a list of
+    # strings instead of a dictionary of keys with empty values.
+    if isinstance(folder_structure, list):
+        assert all(isinstance(item, str) for item in folder_structure)
```
Comment (iLLiCiTiT) [inline client/ayon_core/pipeline/project_folders.py:91]: Please use other exception from `AssertionError` e.g. `ValueError`.

Comment (BigRoy) [reply client/ayon_core/pipeline/project_folders.py]: https://github.com/ynput/ayon-core/pull/1169/commits/ab682da2b83b1e405ed776d6b18ca8234e83d302

File: client/ayon_core/pipeline/project_folders.py [hunk 3]
```diff
@@ -1,6 +1,7 @@
 import os
```
Comment (iLLiCiTiT) [inline client/ayon_core/pipeline/project_folders.py:1]: With this you can replace `List` > `list` and `Dict` > `dict`.
```suggestion
from __future__ import annotations
import os
```

Comment (BigRoy) [reply client/ayon_core/pipeline/project_folders.py:2]: Like this? https://github.com/ynput/ayon-core/pull/1169/commits/8008ab02b543b2d42a0dfbb91bd6afca315d28cf

Comment (iLLiCiTiT) [reply client/ayon_core/pipeline/project_folders.py:2]: Yep
